### PR TITLE
core: add hidden audits for each insight

### PR DESCRIPTION
**Pull request by connorjclark**
ref #16323

* trace engine now exports all translated strings used by each insight
* insight audits mostly just need to use the `adaptInsightToAuditProduct` helper, which is given the insight model and the caller returns a `LH.Audit.Product`
* adds `yarn generate-insight-audits` script, to quickly make and configure new audits (just run when new insights exists, no flags). Calling in CI so we don't forget to add a new insight when upgrading trace engine
* all insight audits are currently hidden until UI is implemented
* only `dom-size-insight`, `viewport-insight`, and `render-blocking-insight` have been implemented. The other 11 are blank implementations
* adds to TraceElements a new group `trace-engine`, which grabs all the node ids referenced in insights and collects their node information

### CODE REVIEW

File: report/renderer/performance-category-renderer.js [hunk 1]
```diff
@@ -201,7 +201,20 @@ export class PerformanceCategoryRenderer extends CategoryRenderer {
       filmstripEl && timelineEl.append(filmstripEl);
     }
 
-    const allInsights = category.auditRefs
+    // Insights
+    const [insightsGroupEl, insightsFooterEl] = this.renderAuditGroup(groups['insights']);
+    insightsGroupEl.classList.add('lh-audit-group--insights');
```
[the remark this answers is not in the document]
Comment (adamraine) [reply report/renderer/performance-category-renderer.js:206]: Hidden audits SGTM. However, we will still need to put some UI logic behind an experimental setting.

I think it makes sense for this PR to establish the structure for insights in the lhr and then a different patch can add the UI logic. That way we can unblock the LH implementations for the remaining insights.

File: core/audits/insights/third-parties-insight.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-unused-vars */ // TODO: remove once implemented.
```
Comment (adamraine) [inline core/audits/insights/third-parties-insight.js:1]: nit: can these move below the `@license`?

File: core/audits/insights/dom-size-insight.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+      /** @type {LH.Audit.Details.Table['headings']} */
+      const headings = [
+        {key: 'statistic', valueType: 'text', label: str_(UIStrings.statistic)},
+        {key: 'node', valueType: 'node', label: str_(UIStrings.element)},
+        {key: 'value', valueType: 'numeric', label: str_(UIStrings.value)},
+      ];
+      /** @type {LH.Audit.Details.Table['items']} */
+      const items = [
+        {
```
Comment (adamraine) [inline core/audits/insights/dom-size-insight.js:52]: Needs total elements #

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -60,6 +60,7 @@ jobs:
     - run: yarn test-legacy-javascript
     - run: yarn i18n:checks
     - run: yarn dogfood-lhci
+    - run: yarn generate-insight-audits
```
Comment (adamraine) [inline .github/workflows/ci.yml:63]: Why does this need to be in our CI? Is it to catch when there's a new insight in the trace engine that we don't have an implementation for?

File: core/gather/gatherers/trace-elements.js [hunk 5]
```diff
@@ -65,6 +65,61 @@ class TraceElements extends BaseGatherer {
...
+            recursiveObjectEnumerate(item, cb, seen);
+          }
+        });
+      }
+    }
+
+    /** @type {number[]} */
+    const nodeIds = [];
+    recursiveObjectEnumerate(insightSet.model, (obj, key) => {
+      if (typeof obj[key] === 'number' && (key === 'nodeId' || key === 'node_id')) {
```
Comment (adamraine) [inline core/gather/gatherers/trace-elements.js:115]: nit: create a `NODE_ID_KEYS` constant in case there are more we need to add to this list

File: core/gather/gatherers/trace-elements.js [hunk 6]
```diff
@@ -65,6 +65,61 @@ class TraceElements extends BaseGatherer {
...
+      return [];
+    }
+
+    /**
+     * Execute `cb(obj, key)` on every object property (non-objects only), recursively.
+     * @param {any} obj
+     * @param {(obj: Record<string, string>, key: string) => void} cb
+     * @param {Set<object>} seen
+     */
+    function recursiveObjectEnumerate(obj, cb, seen) {
```
Comment (adamraine) [inline core/gather/gatherers/trace-elements.js:88]: I think there is some risk that we will resolve node details for a bunch of nodes we don't necessarily need. An insight might hold on to a trace event with a `nodeId` that never gets used in one of the insights.

I guess the alternative is to manually curate the list of `nodeId`s but then we would potentially miss some. False positives > false negatives so this seems good.

File: core/gather/gatherers/trace-elements.js [hunk 7]
```diff
@@ -65,6 +65,61 @@ class TraceElements extends BaseGatherer {
...
+  /**
+   * @param {LH.Artifacts.TraceEngineResult} traceEngineResult
+   * @param {string} mainFrameId
+   * @return {Promise<Array<{nodeId: number}>>}
+   */
+  static async getTraceEngineElements(traceEngineResult, mainFrameId) {
+    // Can only resolve elements for the latest insight set, which should correspond
+    // to the current frame id. Can't resolve elements for pages that are gone.
+    const insightSet = [...traceEngineResult.insights.values()]
+      .reverse().find(insightSet => insightSet.frameId === mainFrameId);
```
Comment (adamraine) [inline core/gather/gatherers/trace-elements.js:77]: Pretty sure this will always be the last insight set in the list. The main frame ID does not change across navigations.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (adamraine) [reply .github/workflows/ci.yml:63]: Ope just read the PR description. SGTM.

File: core/audits/insights/dom-size-insight.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  /**
+   * @param {LH.Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Promise<LH.Audit.Product>}
+   */
+  static async audit(artifacts, context) {
+    return adaptInsightToAuditProduct(artifacts, context, 'DOMSize', (insight) => {
+      if (!insight.maxDOMStats?.args.data.maxChildren || !insight.maxDOMStats?.args.data.maxDepth) {
```
Comment (adamraine) [inline core/audits/insights/dom-size-insight.js:38]: Technically we will still have a valid `totalElements` value even if these are missing. In practice though, this will only happen if the document element is removed so it's an edge case that really doesn't matter.

**Review by adamraine — APPROVED** [review]

File: core/test/scenarios/api-test-pptr.js [hunk 9]
```diff
@@ -104,10 +104,16 @@ describe('Individual modes API', function() {
       } = getAuditsBreakdown(lhr);
       expect(auditResults.map(audit => audit.id).sort()).toMatchSnapshot();
 
-      expect(notApplicableAudits.map(audit => audit.id).sort()).toMatchSnapshot();
+      expect(
+        notApplicableAudits
+          // Flaky in CI.
```
Comment (adamraine) [inline core/test/scenarios/api-test-pptr.js:109]: Seems like a bug, are we tracking it anywhere?

**Review by adamraine — APPROVED** [review]
